### ISSUE TITLE
Issues raised even when "on" threshold

### ISSUE THREAD

**Issue by jotabe87** [detekt/detekt#3679]
Hi everyone! We incorporated detekt in our project a while ago and, while we like it very much, we've been noticing that it seems to report issues even when the checks are "on" the threshold and not after crossing it. 

### Expected Behavior
For this example, imagine any function, with three nested blocks.
```
fun myFunction(): Boolean {
  someNullableVariable?.let {
    if (something) {
      // do something
      if (something else) {
        return true
      }
  }
  return false
}
```

We expect detekt not to report anything, since it's configure like so:
```
  NestedBlockDepth:
    active: true
    threshold: 3
```

### Observed Behavior
The function is reported to have infringed the NestedBlockDepth rule:

`NestedBlockDepth - 3/3 - [applyInternal] at <path>/<file>.kt:16:18`

My team and I believe that this should not be reported, since the number of nested blocks is 3 from the allowed 3 nested blocks. The issue should only be raised, when the set threshold is crossed, for instance at an event where `4/3` could be observed.

## Steps to Reproduce
- Write a similar function 
- Have detekt in your project with the same settings for NestedBlockDepth

## Context
This does not only happen with NestedBlockDepth, but every single rule is treated as such in detekt. While we want to use detekt to ensure code quality, we do not want to have "add one to the real threshold" or suppress the rule in order to get away with it. Also after a while the question could become "Was this a 3 or a 4? The config says 4, so why is 4 not allowed?", mainly if new colleagues come to the project, or the ones that instituted detekt and the rules move to another project.  

## Your Environment
* Version of detekt used: 1.15.0
* Version of Gradle used (if applicable): 6.8
* Operating System and version: macOS Big Sur 11.2.3

**Comment by BraisGabin** [comment]
I do agree that this is missleading but it's impossible to fix right now. We need a major update to make such a change.

**Comment by jotabe87** [comment]
Thanks for the update! I'm looking forward to see this being implemented in the future!

**Comment by cortinico** [comment]
Hey @jotabe87, thanks for reporting. 

> We need a major update to make such a change.

Do we really @BraisGabin? Isn't this considered the same as fixing a false-positive/false-negative bug in a rule? I believe this is fixable provided we add a note in the release notes or so.

**Comment by schalkms** [comment]
This is accepted system behavior. Hence, there's nothing to fix here. This behavior was agreed in the early stages of detekt. Please see #313 for more details.
As of now, this is the first contradicting feedback we got.
I'm opposed to the idea of changing this _again_. There are tools that have one or the other definition of a threshold. Changing the threshold definition requires to change all the default values in the config.yml file. This needs to be done by every single user.

**Comment by BraisGabin** [comment]
I think so. If this problem where only in one rule I'd accept it as kind of false positive. But this fix would affect at least 8 rules. And even with a note in the release notes all our users with custom values there will need to update them. For that reason I think that we should wait for a major.

But I'd like to know the opinion of others because I understand your point.

**Comment by jotabe87** [comment]
@schalkms with all due respect, all I was able to read on [#313](https://github.com/detekt/detekt/issues/313) was a bunch of comments supporting the issue I've raised (the dictionary definition of a threshold), and then the final result being the implementation of a value-inclusive threshold (I believe there was a pool, but I couldn't see the results).

If detekt is settling on this definition, then I'd ask you to communicate it somewhere. I've skimmed the docs a few times now and never got any clarification of what a threshold is in detekt, so I use the most common one when I think of it.

Also about reporting, we've got detekt for about a year in our project now and, though we talked a few times about the issue as a team, we either expected it to be fixed in a coming release or simply never got to actually opening an issue till now. Sometimes these things "bother" people, but they have a lot on their plates and they just don't get to it.

**Comment by BraisGabin** [comment]
I agree with @jotabe87 here.

Not exactly the same but we had a similar problem with `maxIssues`. And there was a comment about it https://github.com/detekt/detekt/pull/2263#issuecomment-575743308 and we fixed it.

For me this is something similar. If I said That I accpet a `NestedBlockDepth` of `1` I expect that all the plain functions pass and any other fail. But in this case I would need to use `2`.

And at work we have this type of configurations:
```yaml
  ComplexInterface:
    active: false
    threshold: 26
  TooManyFunctions:
    active: true
    excludes: ['**/test/**', '**/androidTest/**', '**/commonTest/**', '**/jvmTest/**', '**/jsTest/**', '**/iosTest/**']
    thresholdInFiles: 26
    thresholdInClasses: 26
    thresholdInInterfaces: 26
    thresholdInObjects: 11
    thresholdInEnums: 11
```

`10 + 1` or `25 + 1`. I think that it's easier to think about "the last value that I accept" than "the first number that I reject".

I still think that this is a breaking change and we should not change it until 2.0. But I think that we should change for 2.0.

**Comment by schalkms** [comment]
> with all due respect, all I was able to read on #313 was a bunch of comments supporting the issue I've raised (the dictionary definition of a threshold), and then the final result being the implementation of a value-inclusive threshold (I believe there was a pool, but I couldn't see the results).

There were 4 votes for the current solution and 1 for the one you suggested. Furthermore, all maintainers at this time voted for the currently implemented solution. As a side note, tools like pmd use the same semantic for thresholds as detekt does. It's a decision that needs to be taken at the beginning for obvious reasons.

> If detekt is settling on this definition, then I'd ask you to communicate it somewhere. I've skimmed the docs a few times now and never got any clarification of what a threshold is in detekt, so I use the most common one when I think of it.

What do you miss here? @jotabe87 
https://detekt.github.io/detekt/configurations.html#rule-sets-and-rules

> Also about reporting, we've got detekt for about a year in our project now and, though we talked a few times about the issue as a team, we either expected it to be fixed in a coming release or simply never got to actually opening an issue till now. Sometimes these things "bother" people, but they have a lot on their plates and they just don't get to it.

Let me start to say thanks on behalf of the detekt team for your valuable to time to give feedback in this repo. Your feedback is really appreciated.
Regarding the feedback, the maintainers check multiple communication channels as a source for comments. Sure, not every detekt user can be reached or has the time to submit issues, comments, questions etc.

Alternatively, you could also fork this repo and change the threshold as you like by modifying the source code.

**Comment by schalkms** [comment]
>  But I think that we should change for 2.0.

I don't think so. Let me outline the consequences from a user point of view. If users bump the detekt version, they get a completely different analyzing result with the same detekt configuration. As a user, I'd lose trust in the static analysis tool. As a matter of fact, users have to change every single threshold value in the configuration file and communicate it as well. Moreover, custom detekt rules also have to be adapted to the new behavior.

Why change it, if it's not broken? I don't think there's a pro to change the semantic here. The change requires effort as well.

**Comment by 3flex** [comment]
I never understood the current behaviour either, because I think of a threshold as a limit. But you could also say "it meets the threshold" for reporting. The naming is ambiguous so people have different opinions.

pmd is clearer with the name of the setting. On a nested-if rule the threshold is named "problemDepth" which is less ambiguous to me since it reads to me like "this is the depth at which there's a problem" so I expect a report when that depth is reached.

I'm in favour of changing the status quo. We could do this by introducing new config option names, aliasing and deprecating "threshold" and `ThresholdRule`, so we have a nice migration path. We wouldn't have to wait for version 2.0 to do this.

**Comment by jotabe87** [comment]
> There were 4 votes for the current solution and 1 for the one you suggested. Furthermore, all maintainers at this time voted for the currently implemented solution. As a side note, tools like pmd use the same semantic for thresholds as detekt does. It's a decision that needs to be taken at the beginning for obvious reasons.

Thanks for the clarification @schalkms, I could only read the discussion and couldn't follow the decision process. You're right, it's an important decision that must be made in the very beginning.

> What do you miss here? @jotabe87
https://detekt.github.io/detekt/configurations.html#rule-sets-and-rules

No, I didn't. But I believe it's a problem of semantics. The excerpt reads: _"For example if you want to allow **up to** 20 functions inside a Kotlin file instead of the default threshold of 10, write:"_ and gives the following configuration code

```
complexity:
  TooManyFunctions:
    threshold: 20 
```

The above sentence reads _up to 20_, which means "less than or equal to, but not more than, a stated value, number, level, or time" according do the [dictionary](https://dictionary.cambridge.org/dictionary/english/up-to). In actuality, if I want to allow 20 functions inside a Kotlin file I have to set the threshold to 21. If I set it to 20 as the documentation suggests, I will be allowing for 19 functions to be written in the file. Do you see the confusion here?

> Let me start to say thanks on behalf of the detekt team for your valuable to time to give feedback in this repo. Your feedback is really appreciated.

Thanks, we really like the tool and it's already been very helpful!

> I never understood the current behaviour either, because I think of a threshold as a limit. But you could also say "it meets the threshold" for reporting. The naming is ambiguous so people have different opinions.

I agree with @3flex, the biggest problem here is the ambiguity. So if you decide to change or not to change the way detekt works, either way is fine, but please change the documentation to make it clearer. 

> I'm in favour of changing the status quo. We could do this by introducing new config option names, aliasing and deprecating "threshold" and ThresholdRule, so we have a nice migration path. We wouldn't have to wait for version 2.0 to do this.

Imho I believe it should be changed, even if it's a breaking change. It should ofc be well communicated, with deprecation warnings and "whatnots", so that no one loses trust on the tool when it changes. But that's just my 2¢ as a user.

**Comment by 3flex** [comment]
Commenting to ensure this stays open.

Discussion has stalled, but we need two things to progress:

1. A decision
2. If the decision is to change existing behaviour, then we'll need a plan to implement - this could be by changing behaviour of the existing config options (probably to be done for v2.0, unless we consider this a true bug then we can "fix" in an earlier version) or introducing new config options and deprecating the old. We'll also need a comprehensive list of affected options so we can make the change in a single release.

**Comment by BraisGabin** [comment]
I'll vote to change it. And I think that it should be done in 2.0.

**Comment by atulgpt** [comment]
@BraisGabin @3flex @schalkms I think this can be closed given there is config with the name `threshold`
